api/v1alpha1/gateway_types.go

```diff
@@ -59,9 +59,13 @@ type GatewaySpec struct {
 	// Listeners associated with this Gateway. Listeners define what addresses,
 	// ports, protocols are bound on this Gateway.
 	Listeners []Listener `json:"listeners"`
-	// Routes associated with this Gateway. Routes define
-	// protocol-specific routing to backends (e.g. Services).
-	Routes []core.TypedLocalObjectReference `json:"routes"`
+	// RouteSelector specifies routes to associate with this Gateway. Routes define
+	// protocol-specific routing to back-ends (e.g. Services).
+	//
+	// If undefined, this Gateway will be associated to all routes.
+	//
+	// +optional
+	RouteSelector *RouteSelector `json:"routeSelector,omitempty"`
```

Comment:
How do you define what kind of route to get if we use this approach?

Reply:
@howardjohn all routes will be associated to all gateways unless a `RouteSelector` is specified. A namespace and/or a route may contain one or more labels used for gateway association.

Here's a `LabelSelector` example where routes of two different kinds (`HttpRoute` and `TcpRoute`) are associated to gateway `sample` and `HttpRoute` "foo" is associated to gateway `foo` (no `routeSelector` is defined):
```
apiVersion: networking.x.k8s.io/v1alpha1
kind: Gateway
metadata:
  name: sample
  namespace: gateways
  spec:
    routeSelector:
      labelSelector:
        app: samples
  <SNIP>
---
apiVersion: networking.x.k8s.io/v1alpha1
kind: Gateway
metadata:
  name: foo
  namespace: gateways
  spec:
    <SNIP>
---
apiVersion: networking.x.k8s.io/v1alpha1
kind: HttpRoute
metadata:
  name: sample
  namespace: samples
  labels:
    app: samples
spec:
  <SNIP>
---
apiVersion: networking.x.k8s.io/v1alpha1
kind: TcpRoute
metadata:
  name: sample
  namespace: samples
  labels:
    app: samples
spec:
  <SNIP>
---
apiVersion: networking.x.k8s.io/v1alpha1
kind: HttpRoute
metadata:
  name: foo
  namespace: samples
  labels:
    app: foo
    version: bar
spec:
  <SNIP>
```

Reply:
I am a bit worried about how this would be implemented. routes may select objects other than just HttpRoute and TcpRoute. With this approach, an implementer would need to to query for all objects for the label? I suppose they could limit it to the set of objects that they consider valid routes though

Reply:
> routes may select objects other than just HttpRoute and TcpRoute.

@howardjohn can you help me better understand what you mean here? It's my understanding that routes may be chained together for delegation purposes. In the delegation use case, the labels of the top-level route will be used for gateway association. Take a look at https://github.com/kubernetes-sigs/service-apis/pull/12#issuecomment-571842155 about adding `Routes` to `RouteSelector` to support specific route association.

Thanks for the review and feedback.

Reply:
My understanding of @howardjohn's concern is that TypedLocalObjectReference provides explicit bounds for the set of object types to query, which seems inherently more efficient. A query using an untyped selector is intrinsically unbounded along the object kind dimension.

Does that make sense?

Reply:
The efficiency aspect may become more important if selection across all namespaces is allowed.

Reply:
@ironcladlou as @bowei mentions in https://github.com/kubernetes-sigs/service-apis/pull/12#discussion_r364448727, gateways should be able to bind routes from different namespaces. I'm going to update the `Routes` type to `ObejctReference`.